### PR TITLE
Allow users setup VLAN ID for mgmt interface

### PR DESCRIPTION
**Pull request by tjjh89017**
please refer https://github.com/harvester/harvester/issues/2236

Old config:
```yaml
... omit ...
install:
  ... omit ...
  networks:
    harvester-mgmt:
      interfaces:
      - name: eth0
      - name: eth1
      method: dhcp
    bond-public:
      interfaces:
      - name: eth2
      - name: eth3
      method: none
  ... omit ...
```

New config
```yaml
... omit ...
install:
  ... omit ...
  management_interface: # It will create `br-mgmt.101` as mgmt interface
    interfaces:
    - name: eth0
    - name: eth1
    method: dhcp
    vlan_id: 101
  ... omit ...
# disallow other network interfaces to be configured in the installation stage.
# Installer will have other future PR to support VLAN network pre-configuration.
```

### CODE REVIEW

File: pkg/config/testdata/harvester-config.yaml
Comment (tjjh89017) [reply pkg/config/testdata/harvester-config.yaml:34]: This is for direct config testing only.
I will not add vlan id on it because so far we cannot test vlan id in PXE mode.

File: pkg/console/install_panels.go
Comment (tjjh89017) [reply pkg/console/install_panels.go:902]: I think I should fix it with L810, L812, but not here.
I expected VLAN ID is following `askInterfacePanel`

File: pkg/console/install_panels.go
Comment (tjjh89017) [reply pkg/console/install_panels.go:1082]: I'm not sure how to improve this one.
Could you give me an example for this.
Tks

File: pkg/console/install_panels.go
Comment (tjjh89017) [reply pkg/console/install_panels.go:1585]: I will consider to leave them separately at this moment.
But I found a typo here.
Let me think about it to merge

File: pkg/console/install_panels.go
Comment (tjjh89017) [reply pkg/console/install_panels.go:1082]: I merge it.

File: pkg/console/install_panels.go
Comment (tjjh89017) [reply pkg/console/install_panels.go]: I merge it.

**Comment by tjjh89017** [comment]
I will update PXE example for this PR change once we confirm the break change here.
Thanks a lot

**Comment by bk201** [comment]
This PR also address https://github.com/harvester/harvester/issues/1647

File: pkg/config/config.go
Comment (bk201) [inline pkg/config/config.go:100]: How about naming it `ManagementNetwork`? Since it's a Network type and inside it, there is already a field `Interfaces`.

File: pkg/console/install_panels.go
Comment (bk201) [inline pkg/console/install_panels.go:1088]: By typing an invalid value, user can't see the error.
The installer simply restarts.

File: pkg/config/templates/rke2-canal-config.yaml
Comment (bk201) [inline pkg/config/templates/rke2-canal-config.yaml:9]: Is it OK to remove this HelmChartConfig resource if the iface is empty?

File: pkg/console/validator.go
Comment (bk201) [inline pkg/console/validator.go:186]: Nice to have the check here.

File: pkg/console/install_panels.go
Comment (bk201) [inline pkg/console/install_panels.go:1566]: This is re-used in another place, might be good to implement it as a method in Network.

File: pkg/config/config.go
Comment (tjjh89017) [reply pkg/config/config.go:100]: I'm not sure
ManagementNetwork might conflict with the term of the network of Canal.

File: pkg/console/validator.go
Comment (tjjh89017) [reply pkg/console/validator.go]: fixed

File: pkg/console/install_panels.go
Comment (tjjh89017) [reply pkg/console/install_panels.go]: fixed

File: pkg/console/install_panels.go
Comment (tjjh89017) [reply pkg/console/install_panels.go]: fixed

File: pkg/config/config.go
Comment (bk201) [reply pkg/config/config.go:100]: It's a field name in Harvester configuration; why does it conflict with Canal things?

And do we need to remove the original `Networks` map? We can keep the `Networks` map but only allow the `harvester-mgmt` key here. This provides backward compatibility. For users who don't have other networks configuration, they don't need to alter their existing configuration files.

Comment (tjjh89017) [reply pkg/config/config.go:100]: > Centralized Gateway[¶](https://kubeovn.github.io/docs/v1.10.x/en/guide/subnet/#centralized-gateway)

It will confuse us, we will not sure if we are talking about Pod's Management Network or Management Interface (node's network)

> And do we need to remove the original Networks map? We can keep the Networks map but only allow the harvester-mgmt key here. This provides backward compatibility. For users who don't have other networks configuration, they don't need to alter their existing configuration files.

I will prefer to remove `harvester-mgmt` term in future version.
Make sure it will not confuse anyone.

Comment (bk201) [reply pkg/config/config.go:100]: It will not because [we have the document](https://docs.harvesterhci.io/v1.0/install/harvester-configuration/#installnetworks). We don't present the field to the end-user in the installer UI either. If it confuses anyone, we can add a better document or UI terms.

Harvester is already GA for almost one year, and we already have a user base. So keeping more compatibility here would be appreciated.

Comment (tjjh89017) [reply pkg/config/config.go:100]: > Harvester is already GA for almost one year, and we already have a user base. So keeping more compatibility here would be appreciated.

I didn't really get it about the naming.
If we remove the `harvester-mgmt` and it will never be showed in Linux system.
And if we keep the naming in installer config, won't it confuse the users?

It will only affect in fresh installation but not upgrading.
Nodes from previous version will not be affected at all.

It will be a little bit weird in docs.
we ask users to write config with `harvester-mgmt`, but this keyword doesn't mean anything.

Comment (bk201) [reply pkg/config/config.go:100]: The key here is compatibility and not bringing too much impact to existing users.
I understand we do try our best to migrate existing configurations in the upgrade, but we also need to take care of existing users who have already an infrastructure to use Harvester. For example:
- Customers/users who already have their infrastructure code to spawn Harvester on demand.
- Our QA: they do have infrastructure codes to deploy Harvester repeatedly.
- Our lab maintainer.
- Equinix Metal users.
- Any developers who re-deploy Harvester daily.

Even if we keep the original `Networks`, we still need to check if there are additional configured network keys because they are no longer supported. What I really suggest is to minimize impact. Not arguing the naming.

Comment (tjjh89017) [reply pkg/config/config.go:100]: I will suggest, if we put so many changes in this verison, we should consider to take this chance to migrate all we should.
From my side, I will not consider to leave some legacy entries if we don't need it and that is not for a patch build.

and also you mentioned the additional network keys are not supported anymore, that means some users still need to change the script or be noticed some changes to make sure the scripts will fit the new installer config format.

I also minimize impact for the future change to make sure there is no other will confuse with that entry and code.

But you reminded me, that we should ask users to add `scheme_version` to their config.yaml in the future if we consider to have scheme version for future migration plan.

File: pkg/config/config.go
Comment (guangbochen) [inline pkg/config/config.go:160]: Can u specify what the `SchemeVersion` is used for?

File: pkg/config/config.go
Comment (guangbochen) [reply pkg/config/config.go:100]: both design has pros and cons, if upgrade capability and customer support is our main concerns we may just ask feedbacks internally, and be aware of need to adding those changes to the ipxe repo and existing infra code.

File: pkg/config/cos.go
Comment (guangbochen) [inline pkg/config/cos.go:330]: prefer to return the error in here directly.

File: pkg/config/cos.go
Comment (guangbochen) [inline pkg/config/cos.go:390]:
```suggestion
	postUpScript, err := render("wicked-setup-bond.sh", MgmtBondInterfaceName)
```

File: pkg/config/cos.go
Comment (guangbochen) [inline pkg/config/cos.go:453]: prefer to use `:=` assignment if they are not top level variables.

File: pkg/console/util.go
Comment (guangbochen) [inline pkg/console/util.go:414]: Who will maintains the schemeVersion? if users can modifiy it how can we verify it accoridngly?

File: pkg/config/config.go
Comment (tjjh89017) [reply pkg/config/config.go:160]: Currently, we only logged the version but do nothing about this.
It's reserved for the future.
if we want to do some migration in the future.

File: pkg/config/cos.go
Comment (tjjh89017) [reply pkg/config/cos.go:330]: ok

File: pkg/config/cos.go
Comment (tjjh89017) [reply pkg/config/cos.go:390]: ok

File: pkg/config/cos.go
Comment (tjjh89017) [reply pkg/config/cos.go:453]: ok

File: pkg/console/util.go
Comment (tjjh89017) [reply pkg/console/util.go:414]: Once we change the config struct, we should increase the scheme version.
and users shouldn't change it on their own.
It will be referred to by the future migration script.
Currently, we could not verify it if users change it manually.

File: pkg/config/constants.go
Comment (tjjh89017) [reply pkg/config/constants.go:13]: Canwu asked me to use mgmt-br and mgmt-bo for mgmt interface.
I will check with him again

**Comment by tjjh89017** [comment]
add validater to check if users didn't provide `management_interface` entry and guide users to check docs for new scheme config.

File: pkg/config/constants.go
Comment (yaocw2020) [reply pkg/config/constants.go:13]: The prefix would be the name of the cluster network, thus mgmt is reasonable.

File: pkg/config/config.go
Comment (yaocw2020) [reply pkg/config/config.go:160]: Please add an annotation.

File: pkg/config/cos.go
Comment (yaocw2020) [inline pkg/config/cos.go:315]: What about using if expression rather than switch here because we do nothing in the case branch?

File: pkg/config/cos.go
Comment (yaocw2020) [inline pkg/config/cos.go:331]: The coding mode shown below is more recommend.
``` go
if err := updateBond(stage, MgmtBondInterfaceName, &bondMgmt); err != nil {
    return "", err
}
```

File: pkg/config/cos.go
Comment (yaocw2020) [inline pkg/config/cos.go:345]: What about merging the 315 line switch expression with this switch?

File: pkg/config/cos.go
Comment (yaocw2020) [inline pkg/config/cos.go:357]: I believe `case default:` is necessary for the switch expression.

File: pkg/console/util.go
Comment (yaocw2020) [reply pkg/console/util.go:414]: Please add an annotation.

File: pkg/config/templates/wicked-setup-bridge.sh
Comment (w13915984028) [inline pkg/config/templates/wicked-setup-bridge.sh:14]: `$(ip -json link show dev {{ . }} | jq -j '.[0]["address"]')`

please consider add check and log, to make sure the MAC is really got; and log essential info when failed

File: pkg/console/validator.go
Comment (w13915984028) [inline pkg/console/validator.go:45]: it seems uncommen to add "`" in code for message, please consider to remove them

File: pkg/config/cos.go
Comment (w13915984028) [inline pkg/config/cos.go:437]: it's better to define a const for 4094, it's scattered in the code

File: pkg/config/cos.go
Comment (tjjh89017) [reply pkg/config/cos.go:315]: we will need 3 conditions to match.
but switch-case could use 1 case to cover

File: pkg/config/cos.go
Comment (tjjh89017) [reply pkg/config/cos.go:345]: We need the mgmt name after updateBridge and updateBond.
So we cannot merge it.

File: pkg/config/config.go
Comment (bk201) [reply pkg/config/config.go:100]: The user can know configuration changes:
Screen Shot 2022-09-20 at 2 49 28 PM

For the scheme version part, could we do the same?
Screen Shot 2022-09-20 at 2 46 26 PM

File: pkg/config/cos.go
Comment (tjjh89017) [reply pkg/config/cos.go:357]: We do nothing in this default case.
Just skip it.

File: pkg/config/config.go
Comment (tjjh89017) [reply pkg/config/config.go:100]: fixed

File: pkg/config/templates/wicked-setup-bridge.sh
Comment (tjjh89017) [reply pkg/config/templates/wicked-setup-bridge.sh:14]: move it to bond post up script

File: pkg/console/validator.go
Comment (tjjh89017) [reply pkg/console/validator.go:45]: removed

File: pkg/config/cos.go
Comment (tjjh89017) [reply pkg/config/cos.go:435]: I will not change this in this stage

Comment (tjjh89017) [reply pkg/config/cos.go:435]: There are some ranges between 0 to 4094 and 2 to 4094.
It will confuse if we only set a variable for 4094 but not for 0 and 2.
and it's hard to give a name for 0 and 2.

**Review by yaocw2020 — APPROVED** [review]
LGTM

**Review by w13915984028 — APPROVED** [review]
LGTM, thanks.

**Review by bk201 — APPROVED** [review]
lgtm

**Review by guangbochen — APPROVED** [review]

**Pull request merged**
